Create 2025-03-27-the-case-for-eof.md (#133)
ethereum/solidity-website: src/posts/2025-03-27-the-case-for-eof.md

---
title: 'The Case for EOF'
date: '2025-03-27'
author: Kamil Śliwak
category: Explainers
---

[The EVM Object Format](https://evmobjectformat.org) is a long awaited upgrade that modernizes
the EVM and removes obstacles that have wide-ranging effects on the entire language, tooling and 
application ecosystem on Ethereum.

> Solidity is in full support of the proposal and in this post I would like to explain why.

Reviewer: @mehtavishwa30
> Should we change the narration to "we" of "I"? I prefer it for topics and sentiments that are shared by the team as a whole and not the opinion of a single personal. It bolsters the voice. 

Well, ok. I'll replace "I" with "we". I was actually a bit undecided if I should emphasize some parts as my opinion of just put it from the perspective of the team so I ended up with a mix of both.

I will address the benefits it brings and the reasons why alternatives proposed so far fall short 
of achieving the same goals.

# Benefits of EOF

## Tooling
Development tooling is the area where the benefits of EOF are the most evident.
The biggest flaw of the current EVM in that regard is that the bytecode is not structured enough to be 
effectively analyzed.
This makes developing tools harder and EOF is a significant step towards lowering the entry barrier.

> ### Formal verification and static analysis

Reviewer: @mehtavishwa30 
> Should we mention something about Sourcify's EIP that improves EOF to help improve verification landscape further?
> https://eips.ethereum.org/EIPS/eip-7834
>
> My tweet thread on it:https://x.com/argotorg/status/1878736807766233447

I did link to the article. As for the EIP itself though, I would have to first explain what the problem is and IMO that's going too deep into implementation details.

Tooling and applications will likely follow.

## Applications
dApps will primarily benefit indirectly, from improved compilers and tooling, however, there are
still some smaller directly observable improvements in store for them.

### Relaxing bytecode size limits
One of them is the potential for lifting bytecode size limits ([EIP-7830](https://eips.ethereum.org/EIPS/eip-7830)).
The fact that the size of deployed bytecode cannot exceed ~24 kB has led to multiple complicated
patterns for contract splitting, such as the diamond proxy, which will become unnecessary.

### Gas savings
As our current EOF prototype shows, contracts compiled to EOF are generally a little smaller 
and cheaper.
For example one of the early benchmarks ([Measurable benefits of EOF](https://notes.ethereum.org/@ipsilon/solidity_eof_poc)) indicates an improvement in the range of 10-15% for some real contracts.

While this may seem underwhelming, note that these values come from an incomplete prototype,
with most components of the opcode-based optimizer still unimplemented on EOF.
With block deduplicator, `EXCHANGE` for stack shuffling, tail-call optimization, relaxed inlining, 

> and `RETURNDATACOPY` removal there is potential for higher savings.

Reviewer 2: ```suggestion
and relaxed optimizer restrictions with respect to `RETURNDATACOPY`, there is potential for higher savings.
```

I also want to emphasize that gas savings are not the main point of EOF.
In the design, emphasis was simply on ensuring that existing contracts do not become more expensive, as any increase in cost would have posed a major disincentive to adoption.
EOF already exceeded our expectations in that regard and any gains are just the icing on the top.

## L2s
A common argument is that [layer 2](https://ethereum.org/developers/docs/scaling/) is the place
where innovation like EOF should be taking place since L2s are much more free to experiment with 
the execution layer.
Unfortunately, this not really what is currently happening.
The high value placed on EVM-compatibility means that L2s closely follow L1, and more often lag
behind it than extend it.
It is still the Ethereum mainnet that drives EVM evolution.

We strongly believe that waiting for L2s to take the lead will lead to complete stagnation of the execution layer.
EOF must be first adopted on the mainnet, which will make it very likely to be eventually picked 
up by L2s at their own pace.

And the important aspect is that, like compilers, L2s can be satisfied with a sensible subset

> of the EVM rather than its entirety with all the legacy and quirks, as long as such a subset exists.

Reviewer: @clonker 
> don't understand what you mean here. Why wouldn't a subset exist?

Well, currently it does not. We only have legacy.

But maybe it would be clearer it I added "self-contained" there? It can't be just any random subset :)

Reviewer 2: yeah that's better

They were never adopted, often due to significant and unresolved shortcomings and the
fact that addressing these shortcomings really requires work on the order of what has
been done in EOF.

EOF is well thought out and solves these problems right now, in a single step.
It is extremely well tested compared to other EIPs, having accumulated thousands of
[EEST tests](https://eest.ethereum.org) by now.
It already has a compiler implementation showing that it's feasible.
In theory it can be done differently, but it is not likely that the other approaches,
when perfected, will yield something that is in any way less complex than EOF.

## Can we solve "Stack Too Deep" without EOF?
As Solidity, we have been so vocal about the need to solve the "Stack Too Deep" issue that
one thing needs clarifying.
While it is an important problem, it is largely incidental to EOF and not one of its explicit goals.
[EIP-663](https://eips.ethereum.org/EIPS/eip-663) predates EOF and is simply one of the EIPs
that were always held back the inability to safely add opcodes with immediates.
We asked the Ipsilon team to include it in EOF for three reasons: it depends on it, it is 

> extremely simple and getting it quite directly translates to a lot of saved engineering effort. 

Reviewer: It was originally `getting it quickly` and is literally what I meant. The value is in getting it now, in Fusaka, rather than later.

@clonker @mehtavishwa30 I'd like to change it back, but not sure what the problem was with the original.

---
title: 'The Case for EOF'
date: '2025-03-27'
author: Kamil Śliwak
category: Explainers
---

> [The EVM Object Format](https://evmobjectformat.org) is a long awaited upgrade that modernizes

Reviewer: Well, it's the introduction so it's meant to be vague and the whole post is the elaboration :)

I could change it, but to me it perfectly reflects what it's about, so not even sure how.

the EVM and removes obstacles that have wide-ranging effects on the entire language, tooling and 
application ecosystem on Ethereum.
Solidity is in full support of the proposal and in this post I would like to explain why.
I will address the benefits it brings and the reasons why alternatives proposed so far fall short 

> of achieving the same goals.

Reviewer: Well, it's about EOF's goals, but I'm already repeating EOF way too much.

How about `achieving its goals`?

# Benefits of EOF

## Tooling
Development tooling is the area where the benefits of EOF are the most evident.
The biggest flaw of the current EVM in that regard is that the bytecode is not structured enough to be 
effectively analyzed.
This makes developing tools harder and EOF is a significant step towards lowering the entry barrier.

### Formal verification and static analysis
Exposing more structure is critical for formal verification and static analysis tools, 
which are becoming ever more indispensable.
As soon as the contract gets compiled, it loses a lot useful of high-level information

> that cannot be perfectly reconstructed after the fact.

Reviewer 2: after compilation. like, afterwards the information is lost and cannot 1:1 be reconstructed

> that cannot be perfectly reconstructed after the fact.

Author: yes, but I meant that it doesn't read well so we should complete the sentence. WDYT?

> that cannot be perfectly reconstructed after the fact.

Reviewer: > after what fact?

After the fact that `the contract gets compiled`. What's the proposed change? The suggestion seems empty.

Despite the success of Solidity, this is not something we are happy about.
Having a healthy competition in this space is very important as no language can be one-size-fits-all.

### Why is it so hard?
Why is it so hard for new languages to catch on in this space?
I think it has a lot to do with the EVM itself.
Gas optimization is a major factor.
The existing compilers already have complex optimizers built up and while advanced developers can
still easily beat them with hand-crafted assembly, the fact of the matter is that getting
to an adequate level of optimization in a new compiler is still a big engineering task.

A major example of this is inlining.
Creating good inlining heuristics is a challenge due to two conflicting pressures.
On the one hand, jump instructions are relatively expensive, making it imperative to inline as
much as possible.
On the other hand, inlining functions increases the number of local variables and the required
stack space, so it cannot always be done.
This is a fine line to thread and EOF with its cheaper relative jumps simply provides
much more leeway to a fresh compiler.

The limited stack is another big obstacle that every new project runs into.

> Optimal stack scheduling under constrained access to stack space is an unsolved problem,
> with no optimal algorithm.

Reviewer 2: ```suggestion
Optimal stack scheduling under constrained access to stack space is an unsolved problem.
```

Tooling and applications will likely follow.

## Applications
dApps will primarily benefit indirectly, from improved compilers and tooling, however, there are
still some smaller directly observable improvements in store for them.

### Relaxing bytecode size limits
One of them is the potential for lifting bytecode size limits ([EIP-7830](https://eips.ethereum.org/EIPS/eip-7830)).
The fact that the size of deployed bytecode cannot exceed ~24 kB has led to multiple complicated
patterns for contract splitting, such as the diamond proxy, which will become unnecessary.

### Gas savings
As our current EOF prototype shows, contracts compiled to EOF are generally a little smaller 
and cheaper.
For example one of the early benchmarks ([Measurable benefits of EOF](https://notes.ethereum.org/@ipsilon/solidity_eof_poc)) indicates an improvement in the range of 10-15% for some real contracts.

While this may seem underwhelming, note that these values come from an incomplete prototype,
with most components of the opcode-based optimizer still unimplemented on EOF.
With block deduplicator, `EXCHANGE` for stack shuffling, tail-call optimization, relaxed inlining, 

> and `RETURNDATACOPY` removal there is potential for higher savings.

Reviewer 3: EOF is not removing RETURNDATACOPY, did you mean removal of its OOB exception?

Reviewer 2: it was (probably :)) meant with respect to optimizations that can be performed thanks to relaxed restrictions

Author: Made a change to clarify this. I believe Kamil might have meant that we can remove returndatacopy in some cases (as in https://github.com/ethereum/solidity/issues/15310, "Relax RETURNDATACOPY restrictions in the optimizer").

Reviewer: I'll change it to `and the ability to remove unused `RETURNDATACOPY`. And yeah, it's of course not about removing it from EOF.

These features are not new.
Over the years most of them have been proposed in one way or another.
They were never adopted, often due to significant and unresolved shortcomings and the
fact that addressing these shortcomings really requires work on the order of what has
been done in EOF.

EOF is well thought out and solves these problems right now, in a single step.
It is extremely well tested compared to other EIPs, having accumulated thousands of
[EEST tests](https://eest.ethereum.org) by now.
It already has a compiler implementation showing that it's feasible.
In theory it can be done differently, but it is not likely that the other approaches,
when perfected, will yield something that is in any way less complex than EOF.

## Can we solve "Stack Too Deep" without EOF?
As Solidity, we have been so vocal about the need to solve the "Stack Too Deep" issue that
one thing needs clarifying.
While it is an important problem, it is largely incidental to EOF and not one of its explicit goals.
[EIP-663](https://eips.ethereum.org/EIPS/eip-663) predates EOF and is simply one of the EIPs
that were always held back the inability to safely add opcodes with immediates.

> We asked the Ipsilon team to include it in EOF for three reasons: it depends on it, it is 

Author: what depends on what? EOF depends on EIP-663? Wouldn't they have included it then anyways without us having asked them?

Reviewer: EIP-663 depends on EOF.

value to the stack, we avoid the `JUMPDEST` problem and guarantee that the value is known and constant, just like an immediate. Essentially, we simulate opcodes like this:
```
RJUMP 0x42
```
with this:
```
PUSH1 0x42
RJUMP
```

Hacks of these kind were already proposed in the past and never taken seriously so far.

One of the downsides is that the sequence is 50% longer, which may be an acceptable trade-off for
rarely used opcodes, but would only aggravate the problem if used e.g. with jumps, which are
already overpriced.
It is also much more complicated to implement instructions with a variable number of immediate 
arguments, such as `RJUMPV`.

Overall, while this is a feasible one-off workaround, it is untenable as a general solution to
the problem.
It completely misses the point and aims to patch a single issue in the short-term, while
degrading the design of the EVM in the long term.

# Summary
EOF will benefit the whole stack that depends on the execution layer: compilers, tooling, 
applications, even L2s.
It simplifies the surface of contact with the EVM while also providing the necessary prerequisites 

> to make future extensions less of a struggle.

Author: ```suggestion
to make future extensions easier to implement.
```
